torchmetrics/metric.py

```diff
@@ -221,6 +234,77 @@ def wrapped_func(*args, **kwargs):
+        should_sync: bool = True,
+        distributed_available: Optional[Callable] = distributed_available,
+    ) -> Dict[str, Tensor]:
+        """
+        Sync function for manually controlling when metrics states should be synced across processes
+
+        Args:
+            dist_sync_fn: Function to be used to perform states synchronization
+            process_group:
+                Specify the process group on which synchronization is called.
+                default: None (which selects the entire world)
+            should_sync: Whether to apply to state synchronization. This will have an impact
+                only when running in a distributed setting.
+            distributed_available: Function to determine if we are running inside a distributed setting
+
+        Returns:
+            cache: A dictionary containing the local metric states. The cache will be empty if sync didn't happen.
+        """
+        is_distributed = distributed_available()
+        if not should_sync or not is_distributed:
+            return {}
+        if dist_sync_fn is None:
+            dist_sync_fn = gather_all_tensors
+        # cache prior to syncing
+        cache = {attr: getattr(self, attr) for attr in self._defaults}
+        # sync
+        self._sync_dist(dist_sync_fn, process_group=process_group)
+        return cache
+
+    @contextmanager
+    def sync_context(
+        self,
+        dist_sync_fn: Optional[Callable] = None,
+        process_group: Optional[Any] = None,
+        should_sync: bool = True,
+        restore_cache: bool = True,
+        distributed_available: Optional[Callable] = distributed_available,
+    ) -> None:
+        """
+        Context manager to synchronize the states between processes when running in a distributed setting
+        and restore the local cache states after yielding.
+
+        Args:
+            dist_sync_fn: Function to be used to perform states synchronization
+            process_group:
+                Specify the process group on which synchronization is called.
+                default: None (which selects the entire world)
+            should_sync: Whether to apply to state synchronization. This will have an impact
+                only when running in a distributed setting.
+            restore_cache: Whether to restore the cache state so that the metrics can
+                continue to be accumulated.
+            distributed_available: Function to determine if we are running inside a distributed setting
+        """
+        cache = self.sync(
+            dist_sync_fn=dist_sync_fn,
+            process_group=process_group,
+            should_sync=should_sync,
+            distributed_available=distributed_available
+        )
+
+        yield
+
+        if cache and restore_cache:
+            # if we synced, restore to cache so that we can continue to accumulate un-synced state
+            for attr, val in cache.items():
+                setattr(self, attr, val)
+
```

Comment:
What's the use case for this? If we sync, we should assume all metrics are operating off the synced state and not accumulate local changes, right?

Reply:
This was here already, just moved.

Added in https://github.com/PyTorchLightning/metrics/commit/dd1e744efec0888ddb93abb322d3bc5a4614a540
cc: @SkafteNicki